funflow-aws/src/Control/FunFlow/AWS/S3.hs

```diff
@@ -0,0 +1,97 @@
+{-# LANGUAGE DeriveGeneric         #-}
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE TemplateHaskell       #-}
+{-# LANGUAGE TypeOperators         #-}
+module Control.FunFlow.AWS.S3 where
+
+import qualified Aws
+import qualified Aws.S3                          as S3
+import           Control.FunFlow.Base
+import           Control.FunFlow.ContentHashable
+import           Control.Lens
+import           Control.Monad                   ((>=>))
+import           Control.Monad.Trans.Resource    (runResourceT)
+import           Data.Aeson
+import           Data.Constraint
+import           Data.Reflection
+import qualified Data.Text                       as T
+import           GHC.Generics                    (Generic)
+import           Network.HTTP.Conduit            (newManager,
+                                                  tlsManagerSettings)
+
+-- | Reference to an object in an S3 bucket
+--
+--   Objects can be referenced in a few ways, so this
+--   type is parametrised over the object reference.
+--   Currently, this is expected to be:
+--   - S3.Object (alias for Text)
+--   - S3.ObjectInfo
+data ObjectInBucket obj = ObjectInBucket
+  { _oibBucket :: S3.Bucket
+  , _oibObject :: obj
+  } deriving (Show, Generic)
+
+makeLenses ''ObjectInBucket
+
+instance FromJSON (ObjectInBucket S3.Object)
+instance ToJSON (ObjectInBucket S3.Object)
+
+class ObjectReference a where
+  objectReference :: a -> S3.Object
+
+instance ObjectReference S3.Object where
+  objectReference = id
+
+instance ObjectReference S3.ObjectInfo where
+  objectReference = S3.objectKey
+
+-- | An S3 object is hashable whenever we have sufficient configuration to
+--   access said object. To deal with this, we use reflection to reify a value
+--   (the AWS configuration) into a class constraint.
+--   To use this instance, you must reify the value using 'give':
+--   @
+--     cfg <- Aws.baseConfiguration
+--     give cfg $ contentHash s3object
+--   @
+--
+--   Since S3 is already content hashed, we do not need to actually hash the
+--   object ourselves. In fact, we avoid fetching the object, and only
+--   request the metadata including the content hash.
+--   We incorporate the bucket and name into this to give extra guarantees on
+--   uniqueness, but we may be better abolishing this to deduplicate files
+--   stored in multiple places.
```

Comment:
The content store wouldn't store multiple copies, but dropping the metadata could avoid additional downloads.

Reply:
I meant files stored in multiple places on S3